design-docs/handling-encodings.md

```diff
@@ -0,0 +1,93 @@
+the content (otherwise, the bytes of the source files are copied as is to the destination). This allows filtering the content 
+correctly from and to the file system, as well as from and to archives. 
+
+For the Zip, War, Ear and Tar tasks, we should add support a `fileNameEncoding` encoding option, that should thus be set on their
+parent AbstractArchiveTask.
+
+To be easier to specify in the build file, these properties should be of type `String`, and not of type `Charset`.
+
+## Testing Plan
+
+We need to add tests which cover each of these scenarios:
+
+- The default platform encoding is honored when either of these properties is not set.
+- When `contentEncoding` is set to something other than the default platform encoding, it is honored.
+- When `fileNameEncoding` is set to something other than the default platform encoding, it is honored.
+- When a character cannot be properly converted into the specified encoding, a clear error message is displayed to the user.
+
+## Documentation Plan
+
+- The CopySpec interface documentation should document the `contentEncoding` property and specify the default value if the property is not set.
+  Since the Copy task inherits from CopySpec, that should document the task automatically.
+- The AbstractArchiveTask should document the `fileNameEncoding` property and specify the default value if the property is not set.
+
+## Open Questions
+
+- The Zip, War and Ear tasks are already using `encoding` as the name of its setting which really deals with `fileNameEncoding`. The best option
+  should be to deprecate this property, make the setter set the new `fileNameEncoding` property, and the getter return the `fileNameEncoding` 
+  property.
+- The name `encoding` is much more common practice in things like the xml standards, ant script, editor configurations, etc. 
+  but it's less clear than `contentEncoding`, and is in conflict with the existing `encoding` property of the Zip, War and Ear tasks. Using
+  it would make it impossible to stay backward-compatible.
+- Should the property accept a null value (as it does in the Zip task) and be null by default, and/or should it be set to the default charset?
+- Are there other existing tasks which should deal with character encoding in a better way?
+- Is it a wise choice to use the platform encoding as the default file name encoding? 
+    - The JDK ZipOutputStream defaults to UTF8
+    - Apache comons ZipArchiveOutputStream defaults to UTF8
+    - Apache Ant's ZipOutputStream, which is used by Gradle now, defaults to the platform default encoding, and changing the default would be a  
+      backward incompatibility.
+- Should the Tar task really allow specifying an encoding, and should it default to the platform encoding? It defaults to ASCII, and [the standard](http://www.gnu.org/software/tar/manual/html_node/Standard.html) doesn't seem to allow anything else. 
```

Comment:
@eskatos and I just had a long design discussion inspired by this design document, and we think we've come up with some good, concrete suggestions on how this design should proceed.  I'll start by summarizing the conclusions from that discussion, and then provide some background on how we came up with the names. By the way, even though I'm calling them conclusions, people should definitely point out any errors in our reasoning.

## Conclusions

- `CopySpec` should get a new `filteringCharset` property of type `String` which is the name of the Charset which will be used for encoding and decoding the contents of a file on either side of passing the contents to the closure(s) specified by the various `filter` methods of `CopySpec`
- We should really discuss the creation of archive files separately from extraction from archive files.
- When *creating* Zip archives with the `Zip` task, there should be a `metadataCharset` property which determines the Charset which will be used to create the file entries and header data in the Zip archive.
- When *creating* War, Ear, and Jar files which all extend from the `Zip` task, we should only allow `metadataCharset` to have the value `UTF-8` as that is the expectation of the JVM. We should accomplish this by overriding the setter and getter for the property on the `Jar` task implementation to throw an unsupported operation exception.
- When *creating* a Tar file with the `Tar` task, we should not allow specifying `metadataCharset` using exclusively `UTF-8` and producing POSIX compliant tar files. (Read more above in @eskatos) post.
- When *reading* from Tar files using `TarFileTree` we *should* support a `metadataCharset` option so that we can handle reading Tar files which were created by tools other than Gradle.
- When *reading* from Zip files using `ZipFileTree` we *should* support a `metadataCharset` option as well.
- `filteringCharset` defaults to the JVM default Charset as specified in the `file.encoding` System property.
- `metadataCharset` defaults to `UTF-8` as that is the only allowed value for creating archives other than Zip files in Gradle. (Even though this breaks backwards compatibility, we think it's the right thing to do.)
- Neither property should allow specifying null, and a clear error message should be thrown for any string other than the supported charsets.
- The existing `encoding` property of the `Zip` task should be deprecated and removed in Gradle 3.0
- Finally, we should make a note to go looking through the codebase for properties and method arguments with names like, `encoding`, `charset`, etc, and replace them with either `metadataCharset` or `filteringCharset` if those names are more accurate.

## On Naming

Paul and I first decided that we should switch from names containing `encoding` to `charset` because of a couple of reasons.

1. The user is actually specifying the character set of the bytes as they already exist. That character set will be used both for encoding the bytes into a String and decoding that string back to bytes (in the case of content filtering.)
1. The [TextResourceFactory](https://docs.gradle.org/current/dsl/org.gradle.api.resources.TextResourceFactory.html#org.gradle.api.resources.TextResourceFactory:fromFile(java.lang.Object, java.lang.String)) class also uses `charset` for similar reasons.
1. This is the same name used in several methods in the JVM.

Once we decided that this thing the user is specifying is really the `charset` to be used, we decided that we needed a way to disambiguate whether the `charset` was being used to filter content or to encode/decode the bytes in the metadata of an archive.  Hence the names `filteringCharest` and `metadataCharset`  We also entertained more verbose names like `contentFilteringCharset` and `archiveMetadataCharset` but given the context in which the properties appear to the API user, and the fact that we get to write documentation for the properties, we decided to choose the shorter names. Also, we decided on `filteringCharset` over `contentCharset` because we didn't want to imply that a copySpec with a `contentCharset` specified but *without* specifying filters would actually use the specified charset to write the content of the files being copied to the filesystem.

## Next Steps
@jnizet Please update the design with the changes laid out in this comment.
The rest of the Gradle devs who are interested will give their feedback by next Wednesday, and then we'll consider it designed if there aren't further modifications needed.